omnipaxos_core/tests/reconnect_test.rs

```diff
@@ -0,0 +1,260 @@
+/// * All entries are decided, verify the decided entries
+/// * Only a snapshot was taken, verify the snapshot
+/// * A snapshot was taken and entries decided on afterwards, verify both the snapshot and entries
+fn verify_log_after_recovery(
+    read_log: Vec<LogEntry<Value, LatestValue>>,
+    proposals: Vec<Value>,
+    num_proposals: u64,
+) {
+    match read_log[..] {
+        [LogEntry::Decided(_), ..] => verify_entries(&read_log, &proposals, 0, num_proposals),
+        [LogEntry::Snapshotted(_)] => {
+            let exp_snapshot = LatestValue::create(proposals.as_slice());
+            verify_snapshot(&read_log, num_proposals, &exp_snapshot);
+        }
+        [LogEntry::Snapshotted(_), LogEntry::Decided(_), ..] => {
+            let (first_proposals, last_proposals) =
+                proposals.split_at((num_proposals / 2) as usize);
+            let (first_entry, decided_entries) = read_log.split_at(1); // separate the snapshot from the decided entries
+            let exp_snapshot = LatestValue::create(first_proposals);
+            verify_snapshot(first_entry, num_proposals / 2, &exp_snapshot);
+            verify_entries(decided_entries, last_proposals, 0, num_proposals);
+        }
+        _ => panic!("Unexpected entries in the log: {:?} ", read_log),
+    }
+}
+
+/// Verify that the log has a single snapshot of the latest entry.
+fn verify_snapshot(
+    read_entries: &[LogEntry<Value, LatestValue>],
+    exp_compacted_idx: u64,
+    exp_snapshot: &LatestValue,
+) {
+    assert_eq!(
+        read_entries.len(),
+        1,
+        "Expected snapshot, got: {:?}",
+        read_entries
+    );
+    match read_entries
+        .first()
+        .expect("Expected entry from first element")
+    {
+        LogEntry::Snapshotted(s) => {
+            assert_eq!(s.trimmed_idx, exp_compacted_idx);
+            assert_eq!(&s.snapshot, exp_snapshot);
+        }
+        e => {
+            panic!("{}", format!("Not a snapshot: {:?}", e));
+        }
+    }
+}
+
+/// Verify that all log entries are decided and matches the proposed entries.
+fn verify_entries(
+    read_entries: &[LogEntry<Value, LatestValue>],
+    exp_entries: &[Value],
+    offset: u64,
+    decided_idx: u64,
+) {
+    assert_eq!(
+        read_entries.len(),
+        exp_entries.len(),
+        "read: {:?}, expected: {:?}",
+        read_entries,
+        exp_entries
+    );
+    for (idx, entry) in read_entries.iter().enumerate() {
+        let log_idx = idx as u64 + offset;
+        match entry {
+            LogEntry::Decided(i) if log_idx <= decided_idx => assert_eq!(*i, exp_entries[idx]),
+            LogEntry::Undecided(i) if log_idx > decided_idx => assert_eq!(*i, exp_entries[idx]),
+            e => panic!(
+                "{}",
+                format!(
+                    "Unexpected entry at idx {}: {:?}, decided_idx: {}",
+                    idx, e, decided_idx
+                )
+            ),
+        }
+    }
+}
```

Comment:
I'm pretty sure we have similar functions either in util or consensus_test. Please reuse them or modify them so you don't have to rewrite them here.